internal/pkg/levee/propagation/propagation.go

```diff
@@ -261,6 +241,50 @@ func (prop *Propagation) visitOperands(n ssa.Node, maxInstrReached map[*ssa.Basi
 	}
 }
 
+func (prop *Propagation) visitCall(call *ssa.Call, maxInstrReached map[*ssa.BasicBlock]int, lastBlockVisited *ssa.BasicBlock) {
+	// Some builtins require special handling
+	if builtin, ok := call.Call.Value.(*ssa.Builtin); ok {
+		switch builtin.Name() {
+		// Only the slice being appended to and the referrers can be tainted.
+		// (The elements being appended cannot be tainted.)
+		case "append":
+			prop.visitCallArg(call.Call.Args[0], maxInstrReached, lastBlockVisited)
+			prop.visitReferrers(call, maxInstrReached, lastBlockVisited)
+			return
```

Comment:
`append` does a lot...

* If the capacity of the slice - that is, the size of the underlying array - passed at argument 0 can fit the item(s) to be appended, they are written to the underlying array.
* If the capacity of the slice cannot accommodate the new items, a new array is allocated and elements are copied.
* In either case, the resized (with possibly new-underlying-array) slice is returned.

So strictly speaking, I think `append` should taint only its return argument and the underlying array, not the slice itself.  The arg0 slice only has access to the tainted value if it had capacity and is later re-sliced to expand the capacity.  See this [GoPlayground example](https://play.golang.org/p/SDuDtdz_Te1).

I don't know that we should try to do all that, but it would warrant a comment.

Reply:
The "if there's capacity, the underlying array will be written to" is why I wanted to traverse to the slice that is being appended to (the `arg0` slice).

> I think append should taint only its return argument and the underlying array, not the slice itself.

I agree, but I don't think we can do that. As you mention, we don't want to try modelling the underlying array's capacity. In general, we also don't have a way to get at the underlying array (consider a slice received as a parameter), so I think tainting the slice is a reasonable approximation.

I've added a comment to explain why we want to traverse to the `arg0` slice.